File: minimum-domino-rotations-for-equal-row/minimum-domino-rotations-for-equal-row.java
class Solution {
    public int minDominoRotations(int[] tops, int[] bottoms) {
        int result = Integer.MAX_VALUE;
        for(int target = 1; target < 7;target++){
            int topCount = helper(tops, bottoms, target, true);
            int botCount = helper(tops, bottoms, target, false);
            
            if(topCount != -1 && botCount != -1)
                result = Math.min(result, Math.min(topCount, botCount));
        }
        return result == Integer.MAX_VALUE ? -1 : result;
    }

    int helper(int[] tops, int[] bottoms, int target, boolean topFlag){
        int counter = 0;
        
        for(int i=0;i < tops.length; i++ ){
            if(topFlag && tops[i] != target && bottoms[i] == target )
                counter++;
            else if (!topFlag && tops[i] == target && bottoms[i] != target)
                counter++;
            else if (tops[i] != target && bottoms[i] != target)
                return -1;
        }
        return counter;
    }
    
}

File: 1049-minimum-domino-rotations-for-equal-row/minimum-domino-rotations-for-equal-row.js
/**
 * @param {number[]} tops
 * @param {number[]} bottoms
 * @return {number}
 */
var minDominoRotations = function(tops, bottoms) {
    const n = tops.length;
    
    // Helper function to count rotations needed to make all values equal to target
    const check = (target) => {
        let rotationsTop = 0;
        let rotationsBottom = 0;
        for (let i = 0; i < n; i++) {
            if (tops[i] !== target && bottoms[i] !== target) {
                return -1;
            } else if (tops[i] !== target) {
                rotationsTop++;
            } else if (bottoms[i] !== target) {
                rotationsBottom++;
            }
        }
        return Math.min(rotationsTop, rotationsBottom);
    }

    // Try making all values equal to tops[0] or bottoms[0]
    let rotations = check(tops[0]);
    
    // If tops[0] and bottoms[0] are the same, no need to check again
    if (rotations !== -1 || tops[0] === bottoms[0]) {
        return rotations;
    } else {
        return check(bottoms[0]);
    }
};